### ISSUE TITLE
AttestationInclusion Has Massive Time Complexity

### ISSUE THREAD

**Issue by rauljordan** [prysmaticlabs/prysm#2217]
The `AttestationInclusion` function found [here](https://github.com/prysmaticlabs/prysm/blob/97b5ef0cc2aecd7b356d6dd12129b38bc6e7dedc/beacon-chain/core/balances/rewards_penalties.go#L327) has some of the worst time complexity in our core package. It takes often 3-6 seconds to run with only 48 validators.

The current logic calls the underlying `InclusionSlot` function which is highly inefficient:
```go
for _, index := range prevEpochAttesterIndices {
	slot, err := epoch.InclusionSlot(state, index)
        if err != nil {
		return nil, fmt.Errorf("could not get inclusion slot: %v", err)
	}
}
```

```go
func InclusionSlot(state *pb.BeaconState, validatorIndex uint64) (uint64, error) {
	lowestSlotIncluded := uint64(math.MaxUint64)
	for _, attestation := range state.LatestAttestations {
		participatedValidators, err := helpers.AttestationParticipants(state, attestation.Data, attestation.AggregationBitfield)
		if err != nil {
			return 0, fmt.Errorf("could not get attestation participants: %v", err)
		}
		for _, index := range participatedValidators {
			if index == validatorIndex {
				if attestation.InclusionSlot < lowestSlotIncluded {
					lowestSlotIncluded = attestation.InclusionSlot
				}
			}
		}
	}
	if lowestSlotIncluded == math.MaxUint64 {
		return 0, fmt.Errorf("could not find inclusion slot for validator index %d", validatorIndex)
	}
	return lowestSlotIncluded, nil
}
```

Which calls the extremely inefficient `helpers.AttestationParticipants` function for **every latest attestation in the state** which then calls `CrosslinkCommitteesAtSlot` every single time! We must optimize this.

**Comment by terenc3t** [comment]
this will be addressed in #2151

**Comment by terenc3t** [comment]
Open discussions being held here
https://docs.google.com/document/d/1qYZhIOqXAf_QRPQvrqgaJ2-M_AY1SdWKBmaebjbVOlU/edit#

**Comment by rauljordan** [comment]
Fixed by #2228
